### DISCUSSION TITLE
[ADR] Выбор варианта решения задачи для отображения хинта в компоненте Accordion

### DISCUSSION THREAD

**Discussion by ickisIckis** [zyfra/Prizm#1788] [ADR] [answered]
### Мотив

В процессе обсуждения задачи  #1467  - _Всплывающая подсказка для иконки разворачивания/сворачивания в Accordion - выявили два подхода к реализации_. Необходимо выбрать наиболее удобный для конечных разработчиков вариант решения.

### Краткое описание решения и его обоснование

**Предлагаем два варианта решения:**

1. Template-ориентрированный вариант подразумевает возможность передачи шаблона, который будет использован вместо стандартной кнопки управления аккордеоном. В таком случае, при необходимости добавить хинт использование выглядит так:
``` html
<prizm-accordion>
  <prizm-accordion-item [title]="'My Title'" [expandButton]="expandButtonTemplate">
    <ng-template prizmAccordionContent>
    </ng-template>
  </prizm-accordion-item>
</prizm-accordion>
 <ng-template #expandButtonTemplate let-disabled="disabled">
   <button
      class="accordion__button"
      [class.accordion_expanded__button]="isExpanded"
      [icon]="'chevrons-double-down'"
      [disabled]="disabled"
      [prizmHint]="myHint"
      prizmIconButton
      appearanceType="ghost"
      appearance="secondary"
      size="m"
      tabindex="-1"
  ></button>
</ng-template>
```

**Очевидный плюс этого решения** - _большая гибкость_: разработчики получат возможность управлять любым состоянием кнопки, внешним видом, менять иконки, обрабатывать дополнительные события на кнопке и т.п.

2. Второй подход: добавить хинт для кнопки по умолчанию и возможность заменить текст хинта на стороне проекта.  В таком случае хинт будет доступен всегда, при необходимости замены текстов, это можно будет сделать следующим образом:
```html
<prizm-accordion>
  <prizm-accordion-item [title]="'My Title'" [expandButtonHint]="myHintText$ | async">
    <ng-template prizmAccordionContent>
      // my content
    </ng-template>
  </prizm-accordion-item>
</prizm-accordion>
```

Этот вариант менее гибкий, решает только задачу отображения хинта, но _более лаконичен_ в использовании.

### Последствия
**вариант 1:**
(+)  Большая гибкость решения, решение сразу нескольких задач

(-)  Требуется большее количество кода на стороне проекта

**вариант 2:**
(+)  Лаконичность и простота подхода

(-)  Не гибкое решение, возможно возникновение смежных задач.

**Итог:**
По итогам обсуждения принято решение предоставить обе реализации.
Принято 19.06.2024

**Comment by webatom** [comment, marked as answer]
Вариант "сделать и то и то" не рассматривается?

**Reply by ickisIckis** [reply]
Рассмотрим такой вариант, спасибо!

**Reply by bree7e** [reply]
Нужно сделать первый вариант с default шаблоном. Соответственно если кастомный шаблон не передан, то должен быть отрисован простой вариант.

**Comment by den3011den** [comment]
Если "или-или", то лучше первый. Вопрос не в том понадобится ли эта гибкость, а лишь в том когда пойдут на неë запросы.

**Reply by ickisIckis** [reply]
Да, это верно. Рассмотрим вариант предоставить обе реализации, спасибо.

**Comment by stepandr** [comment]
Тоже конечно за оба варианта с возможностью переключения режима, но в дефолтном исполнении брать второй.

**Reply by ickisIckis** [reply]
спасибо!

**Comment by dkshichkovskiy** [comment]
Второй вариант закрывает текущую потребность на проекте, но первый тоже может быть полезен в перспективе. Поэтому соглашусь с предыдущими комментаторами - полезно иметь оба варианта.

**Reply by ickisIckis** [reply]
спасибо, рассмотрим возможность предоставить оба варианта

**Comment by AleksandrSibiakov** [comment]
Если верить [спеку](https://www.figma.com/design/IEsKtY2w1cx6mvpy3hqB9O/%F0%9F%92%8E-Prizm.-Descriptions-v-1.0?node-id=6640-452101&t=b91ONxtEY9kfy34j-4), то раскрытие происходит по клику на всю верхнюю область, а не только по иконке, таким образом можно предположить что должен быть доступ к некоему toggle API компонента Accordion.
Так же в примерах есть кейсы с тулбарами, отсюда мне напрашивается вариант с полной кастомизацией заголовка, которую можно сделать через компонент [PrizmPanel](https://www.figma.com/design/IEsKtY2w1cx6mvpy3hqB9O/%F0%9F%92%8E-Prizm.-Descriptions-v-1.0?node-id=6242-471605&t=b91ONxtEY9kfy34j-0), где есть возможность задания целой панели инструментов. Для открытия\закрытия аккордиона используем API.

Т.о. кастомизированный аккордион будет выглядеть примерно так:

```
<prizm-accordion>
   <!-- ниже скомбинирована подстановка через директиву-шаблон и PolymorphContent, можно выбрать один, либо поддерживать несколько вариантов (еще добавить ng-content)  -->
  <prizm-accordion-item #accordionItem [title]="example">
    <ng-template #example prizmAccordionHeader> 
       <prizm-panel [title]="'My Title'" (click)="accordionItem.toggle()">
          <ng-container instruments>
             <button
                  [icon]="accordionItem.isExpanded ? 'chevrons-double-down' : 'chevrons-double-up'"                  
                  [prizmHint]="myHint"
                  prizmIconButton
                  appearanceType="ghost"
                  appearance="secondary"
                  size="m"
              ></button>
          </ng-container>
       </prizm-panel>
    </ng-template>

    <ng-template prizmAccordionContent>
      My content
    </ng-template>   
  </prizm-accordion-item>
</prizm-accordion>
```

(P.S. Дефолтную реализацию тоже можно сделать через PrizmPanel, иначе зачем вообще этот компонент существует - везде хедеры переизобретены - сайдбары, аккордионы, Widget и т.д. Я не понимаю, почему не переиспользуется компонент, который даже по дизайн документам [явно указан](https://www.figma.com/design/IEsKtY2w1cx6mvpy3hqB9O/%F0%9F%92%8E-Prizm.-Descriptions-v-1.0?node-id=6448-443128&t=b91ONxtEY9kfy34j-4))

**Reply by ickisIckis** [reply]
сейчас есть задача #1429 как раз о добавлении возможности раскрытия только на кнопку.
рассмотрим в этом контексте задачу тоже. 
По поводу вопроса, почему в текущем аккордеоне не используется панель - это особенности быстрого роста на старте проекта, своеобразный легаси. Спасибо за обратную связь, очень полезно.

**Reply by AleksandrSibiakov** [reply]
Кстати, задача раскрытия только по кнопке тоже решается предложенным подходом: 
```
...
<ng-template #example prizmAccordionHeader> 
       <prizm-panel [title]="'My Title'"> <!-- убираем (click)="accordionItem.toggle()" -->
          <ng-container instruments>
             <button
                  [icon]="accordionItem.isExpanded ? 'chevrons-double-down' : 'chevrons-double-up'"                  
                  [prizmHint]="myHint"
                  prizmIconButton
                  appearanceType="ghost"
                  appearance="secondary"
                  size="m"
                  (click)="accordionItem.toggle()" <-- добавляем на кнопку
              ></button>
          </ng-container>
       </prizm-panel>
       ...
```

Остается только вопрос, что должно быть дефолтным.

**Comment by AleksandrSibiakov** [comment]
Если рассматривать отдельно второй вариант, мне кажется имеет смысл попробовать добиться нужного поведения через токены локализации - иметь дефолтную подсказку библиотека уже должна (как для любого интерактивного элемента), а сменить ее текст можно подменив переводы в скоупе компонента или инстанса Accordion (через директиву с кастомными провайдерами локализации, например)

**Reply by ickisIckis** [reply]
Это базовая часть варианта 2: дефолтный текст через токен локализации. Но по отзывам пользователей не всем удобно использовать токены, часто просят добавить возможность прокинуть свой текст на месте: подобные запросы точно были в кроне и file upload.

**Reply by AleksandrSibiakov** [reply]
Может попытаться улучшить опыт работы с локализацией? Я с такими задачами пока не сталкивался в рамках PRIZM, но мне кажется расширять сам компонент спец. инпутами не очень красиво, можно вынести эту логику в отдельные директивы - либо единая под все, с возможностью частично переопределить (либо несколько точечных директив под конкретные элементы):

```
@Directive({
  selector: prizm-accordion[prizmI18n],
  providers: [
   { 
     provide: PrizmI18nToken,
     useFactory: () => new Proxy(inject(PrizmI18nToken, { skipSelf: true }), { get: (prop, target) => inject(PrizmAccordionL10nDirective, { self: true }).custom?[prop] ?? target[prop] })
  ]
}) 
PrizmAccordionL10nDirective {
    @Input('prizmI18n') custom: Record<string, string>;
}
```

\+ чистый компонент
\- не слишком прозрачный набор доступных возможностей (в помощь документация и листинг IDE (aka intellisense))

**Reply by ickisIckis** [reply]
Заведите issue ;)

**Reply by AleksandrSibiakov** [reply]
Так вот же RFC ? ))

Как я писал выше - проблем с локализацией не испытываю - написать директиву по необходимости, я считаю, разработчик в силах и на стороне проекта. 
Просто предложил как вариант, раз такой вопрос поднялся на уровне библиотеки

**Comment by AsyaSokolova** [comment]
Думаю первый вариант лучше, такой подход уже используется в других компонентах, он удобен и дает больше гибкости
